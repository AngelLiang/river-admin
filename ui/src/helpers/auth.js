

import http from "@/helpers/http"

export const ADD = "add"
export const CHANGE = "change"
export const DELETE = "delete"
export const VIEW = "view"

export const WORKFLOW = "workflow"
export const STATE = "state"
export const FUNCTION = "function"
export const TRANSITION_META = "transitionmeta"
export const TRANSITION_APPROVAL_META = "transitionapprovalmeta"
export const TRANSITION_HOOK = "ontransithook"
export const APPROVAL_HOOK = "onapprovedhook"
export const TRANSITION = "transition"
export const TRANSITION_APPROVAL = "transitionapproval"



class Auth {

    _has_permission(operation, object_type, callback) {
        return http.get(`/user/has_river_permission/${operation}/${object_type}`, response => {
            if (response.data) {
                callback(true)
            } else {
                callback(false)
            }
        })
    }

    has_add_permission(object_type, callback) {
        return this._has_permission(ADD, object_type, callback)
    }

    has_change_permission(object_type, callback) {
        return this._has_permission(CHANGE, object_type, callback)
    }

    has_delete_permission(object_type, callback) {
        return this._has_permission(DELETE, object_type, callback)
    }

    has_view_permission(object_type, callback) {
        return this._has_permission(VIEW, object_type, callback)
    }
}

export var auth = new Auth();