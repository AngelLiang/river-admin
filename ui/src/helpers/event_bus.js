import Vue from 'vue';
const EventBus = new Vue();

const SUCCESS_EVENT_NAME = "SUCESS";
const ERROR_EVENT_NAME = "ERROR";
const LOGOUT_EVENT_NAME = "LOGOUT"

export function emit_success(payload, timeout) {
    EventBus.$emit(SUCCESS_EVENT_NAME, payload, timeout)
}

export function emit_error(payload, timeout) {
    EventBus.$emit(ERROR_EVENT_NAME, payload, timeout)
}

export function emit_logout() {
    EventBus.$emit(LOGOUT_EVENT_NAME)
}

export function on_success(callback) {
    EventBus.$on(SUCCESS_EVENT_NAME, callback)
}

export function on_error(callback) {
    EventBus.$on(ERROR_EVENT_NAME, callback)
}

export function on_logout(callback) {
    EventBus.$on(LOGOUT_EVENT_NAME, callback)
}