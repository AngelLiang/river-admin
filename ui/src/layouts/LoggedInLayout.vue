<template>
  <v-app id="inspire" v-if="initialized">
    <v-app-bar @click.stop="mini = true" app clipped-right color="blue" dark>
      <v-toolbar-title :style="{ cursor: 'pointer'}" @click="goToHomePage">
        <v-row justify="center" align="center">
          <v-col>
            <img class="mt-2 toolbar-logo" src="@/assets/logo.svg" />
          </v-col>
          <v-col>
            <span class="app-title">River Admin</span>
          </v-col>
        </v-row>
      </v-toolbar-title>
      <div class="flex-grow-1"></div>

      <v-tooltip bottom>
        <template v-slot:activator="{ on }">
          <v-icon class="mr-2" v-on="on">mdi-account</v-icon>
        </template>
        <span>{{ user_profile.username }}</span>
      </v-tooltip>
      <v-btn icon @click="logout">
        <v-icon>mdi-logout</v-icon>
      </v-btn>
    </v-app-bar>

    <v-content>
      <v-container align="start" justify="start" class="pa-0" fluid>
        <v-row no-gutters align="start" justify="start">
          <v-col xl="auto" lg="auto" md="auto" sm="auto" xs="auto">
            <div :class="mini?'mini-drawer-shadow':'drawer-shadow'" />
            <v-navigation-drawer
              v-model="drawer"
              :mini-variant="mini"
              permanent
              fixed
              class="drawer"
            >
              <v-list-item jusitfy="center" align="center">
                <v-list-item-title>Menu</v-list-item-title>
              </v-list-item>

              <v-divider />

              <v-list dense>
                <v-list-item
                  :disabled="item.disabled"
                  :id="item.id"
                  v-for="item in items"
                  :key="item.title"
                  @click="goTo(item)"
                  link
                >
                  <v-list-item-icon>
                    <v-icon :disabled="item.disabled">{{ item.icon }}</v-icon>
                  </v-list-item-icon>

                  <v-list-item-content>
                    <v-list-item-title>{{ item.title }}</v-list-item-title>
                  </v-list-item-content>
                </v-list-item>
              </v-list>

              <v-divider />

              <v-list dense>
                <v-list-item
                  :id="item.id"
                  v-for="item in workflow_items"
                  :key="item.title"
                  @click="goTo(item)"
                  link
                >
                  <v-list-item-icon>
                    <v-icon>{{ item.icon }}</v-icon>
                  </v-list-item-icon>

                  <v-list-item-content>
                    <v-list-item-title>{{ item.title }}</v-list-item-title>
                  </v-list-item-content>
                </v-list-item>
              </v-list>
              <v-divider />

              <v-divider />
              <v-row>
                <div class="flex-grow-1"></div>
                <v-col>
                  <v-btn icon @click.stop="mini = !mini">
                    <v-icon v-if="mini">mdi-chevron-double-right</v-icon>
                    <v-icon v-else>mdi-chevron-double-left</v-icon>
                  </v-btn>
                </v-col>
                <div class="flex-grow-1"></div>
              </v-row>
            </v-navigation-drawer>
          </v-col>
          <v-col>
            <slot></slot>
          </v-col>
        </v-row>
      </v-container>
    </v-content>
  </v-app>
</template>

<script>
import { emit_logout } from "@/helpers/event_bus";
import { auth, WORKFLOW, FUNCTION, STATE } from "@/helpers/auth";
import http from "@/helpers/http";

export default {
  name: "LoggedInLayout",
  data: () => ({
    initialized: false,
    drawer: false,
    mini: true,
    items: [],
    workflow_items: [],
    user_profile: null,
    raw_items: [
      { id: "workflow-list", title: "Workflows", icon: "mdi-file-tree", name: "list-workflows", authorization_object_type: WORKFLOW },
      { id: "state-list", title: "States", icon: "mdi-label-variant-outline", name: "list-states", authorization_object_type: STATE },
      { id: "function-list", title: "Functions", icon: "mdi-function-variant", name: "list-callback-functions", authorization_object_type: FUNCTION }
    ]
  }),
  mounted() {
    if (!this.initialized) {
      var user_profile_fetcher = http.get("/user/get/", response => {
        this.user_profile = response.data;
      });

      var workflow_fetchers = auth.has_view_permission(WORKFLOW, yes => {
        if (yes) {
          return http.get("/workflow/metadata", response => {
            this.workflow_items = response.data.map(workflow_metadata => ({
              id: workflow_metadata.id,
              title: workflow_metadata.name,
              icon: workflow_metadata.icon,
              name: "list-workflow-objects",
              params: { workflow_id: workflow_metadata.id }
            }));
          });
        }
      });

      Promise.all(
        this.raw_items
          .map((raw_item, index) => {
            if (raw_item.authorization_object_type) {
              return auth.has_view_permission(raw_item.authorization_object_type, yes => {
                this.items.push({ id: raw_item.id, title: raw_item.title, icon: raw_item.icon, name: raw_item.name, index, disabled: !yes });
              });
            }
          })
          .concat([user_profile_fetcher, workflow_fetchers])
      ).then(() => {
        this.items = this.items.sort((i1, i2) => i1.index - i2.index);
        this.initialized = true;
      });
    }
  },
  methods: {
    logout() {
      emit_logout();
    },
    goToHomePage() {
      this.$router.push({ name: "home" });
    },
    goTo(item) {
      this.$router.push({ name: item.name, params: item.params || {} });
    }
  }
};
</script>

<style>
.toolbar-logo {
  width: 45px;
}

.drawer {
  top: 65px !important;
}

.drawer-shadow {
  width: 250px !important;
}

.mini-drawer-shadow {
  width: 80px;
}

.app-title {
  font-weight: bolder;
}
</style>