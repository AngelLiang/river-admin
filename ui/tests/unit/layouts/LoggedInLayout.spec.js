
import Vuetify from 'vuetify'
import Vue from 'vue'
import { auth, WORKFLOW, STATE, VIEW, FUNCTION } from "../../../src/helpers/auth"

import http from "../../../src/helpers/http"
import { shallowMount } from '@vue/test-utils'
import LoggedInLayout from '@/layouts/LoggedInLayout.vue'

describe('LoggedInLayout.vue', () => {
    let vuetify

    beforeEach(() => {
        vuetify = new Vuetify()
        Vue.use(Vuetify)
    })
    it('Should render workflow list on the left panel if the user is authorized', async () => {
        var has_permission_spy = jest.spyOn(auth, "_has_permission").mockImplementation(
            (operation, object_type, callback) =>
                (Promise.resolve((operation == VIEW && object_type == WORKFLOW) ? callback(true) : callback(false)))
        )

        var workflows_spy = jest.spyOn(http, "get").mockImplementation(
            (url, callback) =>
                (Promise.resolve((url == "/user/get/") ? callback({ data: { username: "testuser" } }) : callback({ data: [] })))
        );


        const wrapper = shallowMount(LoggedInLayout, {
            vuetify,
            sync: false,
        })



        await wrapper.vm.$nextTick();


        expect(wrapper.vm.items).toContainEqual(expect.objectContaining({ id: "workflow-list", disabled: false }))

        expect(wrapper.element).toMatchSnapshot()

        has_permission_spy.mockRestore()
        workflows_spy.mockRestore()
    })

    it('Should render workflow list as disabled on the left panel if the user is not authorized', async () => {
        var has_permission_spy = jest.spyOn(auth, "_has_permission").mockImplementation(
            (operation, object_type, callback) =>
                (Promise.resolve(callback(false)))
        )

        var workflows_spy = jest.spyOn(http, "get").mockImplementation(
            (url, callback) =>
                (Promise.resolve((url == "/user/get/") ? callback({ data: { username: "testuser" } }) : callback({ data: [] })))
        );


        const wrapper = shallowMount(LoggedInLayout, {
            vuetify,
            sync: false,
        })



        await wrapper.vm.$nextTick();


        expect(wrapper.vm.items).not.toContainEqual(expect.objectContaining({ id: "workflow-list", disabled: false }))

        expect(wrapper.element).toMatchSnapshot()

        has_permission_spy.mockRestore()
        workflows_spy.mockRestore()
    })

    it('Should render state list on the left panel if the user is authorized', async () => {
        var has_permission_spy = jest.spyOn(auth, "_has_permission").mockImplementation(
            (operation, object_type, callback) =>
                (Promise.resolve((operation == VIEW && object_type == STATE) ? callback(true) : callback(false)))
        )

        var workflows_spy = jest.spyOn(http, "get").mockImplementation(
            (url, callback) =>
                (Promise.resolve((url == "/user/get/") ? callback({ data: { username: "testuser" } }) : callback({ data: [] })))
        );


        const wrapper = shallowMount(LoggedInLayout, {
            vuetify,
            sync: false,
        })



        await wrapper.vm.$nextTick();


        expect(wrapper.vm.items).toContainEqual(expect.objectContaining({ id: "state-list", disabled: false }))

        expect(wrapper.element).toMatchSnapshot()

        has_permission_spy.mockRestore()
        workflows_spy.mockRestore()
    })

    it('Should render state list as disabled on the left panel if the user is not authorized', async () => {
        var has_permission_spy = jest.spyOn(auth, "_has_permission").mockImplementation(
            (operation, object_type, callback) =>
                (Promise.resolve(callback(false)))
        )

        var workflows_spy = jest.spyOn(http, "get").mockImplementation(
            (url, callback) =>
                (Promise.resolve((url == "/user/get/") ? callback({ data: { username: "testuser" } }) : callback({ data: [] })))
        );


        const wrapper = shallowMount(LoggedInLayout, {
            vuetify,
            sync: false,
        })



        await wrapper.vm.$nextTick();


        expect(wrapper.vm.items).not.toContainEqual(expect.objectContaining({ id: "state-list", disabled: false }))

        expect(wrapper.element).toMatchSnapshot()

        has_permission_spy.mockRestore()
        workflows_spy.mockRestore()
    })

    it('Should render function list on the left panel if the user is authorized', async () => {
        var has_permission_spy = jest.spyOn(auth, "_has_permission").mockImplementation(
            (operation, object_type, callback) =>
                (Promise.resolve((operation == VIEW && object_type == FUNCTION) ? callback(true) : callback(false)))
        )

        var workflows_spy = jest.spyOn(http, "get").mockImplementation(
            (url, callback) =>
                (Promise.resolve((url == "/user/get/") ? callback({ data: { username: "testuser" } }) : callback({ data: [] })))
        );


        const wrapper = shallowMount(LoggedInLayout, {
            vuetify,
            sync: false,
        })



        await wrapper.vm.$nextTick();

        expect(wrapper.vm.items).toContainEqual(expect.objectContaining({ id: "function-list" }))

        expect(wrapper.element).toMatchSnapshot()

        has_permission_spy.mockRestore()
        workflows_spy.mockRestore()
    })

    it('Should render state list as disabled on the left panel if the user is not authorized', async () => {
        var has_permission_spy = jest.spyOn(auth, "_has_permission").mockImplementation(
            (operation, object_type, callback) =>
                (Promise.resolve(callback(false)))
        )

        var workflows_spy = jest.spyOn(http, "get").mockImplementation(
            (url, callback) =>
                (Promise.resolve((url == "/user/get/") ? callback({ data: { username: "testuser" } }) : callback({ data: [] })))
        );


        const wrapper = shallowMount(LoggedInLayout, {
            vuetify,
            sync: false,
        })



        await wrapper.vm.$nextTick();


        expect(wrapper.vm.items).not.toContainEqual(expect.objectContaining({ id: "function-list", disabled: false }))

        expect(wrapper.element).toMatchSnapshot()

        has_permission_spy.mockRestore()
        workflows_spy.mockRestore()
    })

    it('Should render workflow object list on the left panel if the user is authorized', async () => {
        var has_permission_spy = jest.spyOn(auth, "_has_permission").mockImplementation(
            (operation, object_type, callback) =>
                (Promise.resolve((operation == VIEW && object_type == WORKFLOW) ? callback(true) : callback(false)))
        )

        var workflows_spy = jest.spyOn(http, "get").mockImplementation(
            (url, callback) =>
                (Promise.resolve((url == "/user/get/") ? callback({ data: { username: "testuser" } }) :
                    ((url == "/workflow/metadata") ? callback({ data: [{ id: 1, name: "Transportation Flow", icon: "mdi-truck" }] })
                        : callback({ data: [] }))))
        );


        const wrapper = shallowMount(LoggedInLayout, {
            vuetify,
            sync: false,
        })



        await wrapper.vm.$nextTick();


        expect(wrapper.vm.workflow_items).toContainEqual(expect.objectContaining({ id: 1, title: "Transportation Flow" }))

        expect(wrapper.element).toMatchSnapshot()

        has_permission_spy.mockRestore()
        workflows_spy.mockRestore()
    })


    it('Should not render state list on the left panel if the user is not authorized', async () => {
        var has_permission_spy = jest.spyOn(auth, "_has_permission").mockImplementation(
            (operation, object_type, callback) =>
                (Promise.resolve(callback(false)))
        )

        var workflows_spy = jest.spyOn(http, "get").mockImplementation(
            (url, callback) =>
                (Promise.resolve((url == "/user/get/") ? callback({ data: { username: "testuser" } }) : callback({ data: [] })))
        );


        const wrapper = shallowMount(LoggedInLayout, {
            vuetify,
            sync: false,
        })



        await wrapper.vm.$nextTick();


        expect(wrapper.vm.workflow_items).not.toContainEqual(expect.objectContaining({ id: 1, title: "Transportation Flow" }))

        expect(wrapper.element).toMatchSnapshot()

        has_permission_spy.mockRestore()
        workflows_spy.mockRestore()
    })
})