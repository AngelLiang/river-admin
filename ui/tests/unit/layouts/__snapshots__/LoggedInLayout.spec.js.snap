// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoggedInLayout.vue Should not render state list on the left panel if the user is not authorized 1`] = `<!---->`;

exports[`LoggedInLayout.vue Should render function list on the left panel if the user is authorized 1`] = `<!---->`;

exports[`LoggedInLayout.vue Should render state list as disabled on the left panel if the user is not authorized 1`] = `<!---->`;

exports[`LoggedInLayout.vue Should render state list as disabled on the left panel if the user is not authorized 2`] = `<!---->`;

exports[`LoggedInLayout.vue Should render state list on the left panel if the user is authorized 1`] = `<!---->`;

exports[`LoggedInLayout.vue Should render workflow list as disabled on the left panel if the user is not authorized 1`] = `<!---->`;

exports[`LoggedInLayout.vue Should render workflow list on the left panel if the user is authorized 1`] = `<!---->`;

exports[`LoggedInLayout.vue Should render workflow object list on the left panel if the user is authorized 1`] = `<!---->`;
